netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -280,10 +283,29 @@ protected boolean isGracefulShutdownComplete() {
    * the creation request is queued.
    */
   private void createStream(CreateStreamCommand command, final ChannelPromise promise) {
-    final int streamId = getAndIncrementNextStreamId();
+    // Get the stream ID for the new stream.
+    final int streamId;
+    try {
+      streamId = getAndIncrementNextStreamId();
+    } catch (StatusException e) {
+      // Stream IDs have been exhausted for this connection. Fail the promise immediately.
+      promise.setFailure(e);
+
+      // If we haven't already started a graceful shutdown of the connection, do so now.
+      if (ctx.channel().isActive() && !connection().goAwaySent()) {
```

Comment:
The check for `goAwaySent()` looks like something that `Http2ConnectionHandler` should do. It isn't today, so we will need this here, but it may be good to add to Netty. At that point we wouldn't need this condition at all.